File: partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum=0,ans=0;
        for(int i=0;i<nums.size();i++)
        {
            sum+=nums[i];

        }
        if(sum%2!=0)
            return false;
        else 
            ans=sum/2;
        int n=nums.size();
        bool t[n+1][ans+1];
        
        for(int j=0;j<ans+1;j++)
        {
            t[0][j]=false;
        }
        
        for(int i=0;i<n+1;i++)
        {
            t[i][0]=true;
        }
        
        
        for(int i=1;i<n+1;i++)
        {
            for(int j=1;j<ans+1;j++)
            {
                if(j<nums[i-1])
                    t[i][j]=t[i-1][j];
                else if (j>=nums[i-1])
                    t[i][j]=t[i-1][j]||t[i-1][j-nums[i-1]];
                
            }
        }
        return t[n][ans];
        
        
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        dp = [[-1]*(s//2+1) for _ in range(len(nums)+1)]
        if s%2==1:
            return False
        def solve(ind, su):
            if su == s//2:
                return True
            if ind>=len(nums) or su>s//2:
                return False
            if dp[ind][su] != -1:
                return dp[ind][su]
            if su + nums[ind]<=s//2:
                dp[ind][su] = solve(ind+1, su+nums[ind]) or solve(ind+1, su)
            else:
                dp[ind][su] = solve(ind+1, su)
            return dp[ind][su]
        return solve(0,0)